hedera-node/src/main/java/com/hedera/services/records/ConsensusTimeTracker.java

```diff
@@ -0,0 +1,312 @@
+package com.hedera.services.records;
+/*
+ * ‌
+ * Hedera Services Node
+ * ​
+ * Copyright (C) 2018 - 2022 Hedera Hashgraph, LLC
+ * ​
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	 http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+import java.time.Instant;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.swirlds.platform.state.TransactionHandler;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Provides a tracker of the usable consensus time space during a
+ * {@link com.hedera.services.state.logic.StandardProcessLogic#incorporateConsensusTxn} call.
+ */
+@Singleton
+public class ConsensusTimeTracker {
+	private static final Logger log = LogManager.getLogger(ConsensusTimeTracker.class);
+
+	@VisibleForTesting
+	static final long DEFAULT_NANOS_PER_ROUND = TransactionHandler.MIN_TRANS_TIMESTAMP_INCR_NANOS;
+
+	/**
+	 * The max number of records that can be placed before the first transaction in a round. We need this because
+	 * the first transaction might have more than a normal number of proceeding records.
```

Comment:
Typo I think, should be "preceding records".
```suggestion
	 * the first transaction might have more than a normal number of preceding records.
```

Reply:
The problem then is we need to enforce that this can only happen at the _first_ transaction that the system processes after startup. Otherwise there is no way to choose how much space to leave between transactions for any possible preceding records.

Reply:
I made it so on the first startup you can have an unlimited number of "proceeding" records.